### ISSUE TITLE
Nitrokey HSM 2 don't work properly under WIN

### ISSUE THREAD

**Issue by mti-sk** [OpenSC/OpenSC#1861]
### Problem Description
 I try login to web (IE9) with private certificate on Nitrokey HSM 2, but if I use OpenSC version 0.19.0x64 i get message: "Nitrokey Nitrokey HSM SmartCard reader error", if I use latest OpenSC 0.20.0-rc2(x64) - same messge
Windows & IE9 can't use  nitrokey reader HSM.

Note:
reg: 
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\SmartCard-HSM 4K]
"80000001"="C:\Program Files\OpenSC Project\OpenSC\minidriver\opensc-minidriver.dll"
"ATR"=hex:3b,de,18,ff,81,91,fe,1f,c3,80,31,81,54,48,53,4d,31,73,80,21,40,81,07,1c

### Steps to reproduce
Nitrokey HSM 2 (with private certificate)
Internet Expolorer 9 
Login to web page via certificate on token

### Log
nothing bad or error in md.log

==================================================================

P:1556 T:11824 pCardData:000001BA15C4D860 CardAcquireContext, dwVersion=7, name=SmartCard-HSM 4K,hScard=0x00000000, hSCardCtx=0x00000004
request version pCardData->dwVersion = 7
pCardData->dwVersion = 7
create sc ccontext
sc context created
associate_card
sc_ctx_get_reader_count(ctx): 1
Connected card in 'Nitrokey Nitrokey HSM 0'
PKCS#15 initialization result: 0, Success

...

P:1556 T:11824 pCardData:000001BA15C7A7B0 CardReadFile
pszDirectoryName = mscp, pszFileName = cmapfile, dwFlags = 0, pcbData=0000000404BFE7F0, ppbData=0000000404BFE7F8
check_reader_status
pCardData->hSCardCtx:0x00000004 hScard:0x00000000
check_reader_status r=5 flags 0x00000005
MD virtual file system: found 'mscp' directory
MD virtual file system: found 'cmapfile' file
returns 'cmapfile' content:
--- 000001BA15CDB780:1032
 0000  43006500 72007400 69006600 69006300  61007400 65000000 00000000 00000000
...
 0400  00000000 00000000 

P:1556 T:11824 pCardData:000001BA15C7A7B0 CardGetProperty('PIN Information',cbData=36,dwFlags=1) called
check_reader_status
pCardData->hSCardCtx:0x00000004 hScard:0x00000000
check_reader_status r=5 flags 0x00000005
returning info on normal PIN [1]
returns 'PIN Information' --- 000001BA15CD6830:36
 0000  06000000 00000000 05000000 02000000  04000000 06000000 00000000 00000000
 0020  00000000 

P:1556 T:11824 pCardData:000001BA15C7A7B0 CardReadFile
pszDirectoryName = mscp, pszFileName = msroots, dwFlags = 0, pcbData=0000000404BFF0C0, ppbData=0000000404BFF0C8
check_reader_status
pCardData->hSCardCtx:0x00000004 hScard:0x00000000
check_reader_status r=5 flags 0x00000005
MD virtual file system: found 'mscp' directory
CardReadFile(): file 'msroots' not found in 'mscp'

P:1556 T:11824 pCardData:000001BA15C7A7B0 hScard=0x00000000 hSCardCtx=0x00000004 CardDeleteContext
disassociate_card
sc_pkcs15_unbind
sc_disconnect_card
release context
**********************************************************************

**Comment by dengert** [comment]
You are trying to use a `Nitrokey HSM 2` Deos it have the same ATR as "SmartCard-HSM 4K"?
ANd you alony printed part of the registry entry. 

Why is certificate private? Every card I have ever seen has readable certificates. Windows makes the assumption that all certificates are readable. This allows windows to save certificates in the certificate store, so it can find a certificate then ask user to insert the correct smart card.

**Comment by mti-sk** [comment]
Hi,
My mistake, I don't really distinguish between certificates and keys. The certificate is public and the key is not. And so in this case, the certificates are readable.
But back: The main problem is that Windows reports: smartcard reader error, this smart card needs to install driver drivers that are not on this system!

I don't know why nitrokey HSM 2 and SmardCard HSM 4K have the same ATR. Maybe this is because Nitrokey uses SmartCard.de technology. I have no idea how ATR manufacturers are acquiring or allocating it (I think it will not be a completely clean solution).

**Comment by dengert** [comment]
Above you listed the registry, but it is not complete. It is missing "Crypto Provider" and "Smart Card Key Storage Provider" So "basecsp.dll' is never found. 
 It should look more like this one that is installed by 0.20.0-rc3. 
```
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\SmartCard-HSM 4K]
"Crypto Provider"="OpenSC CSP"
"Smart Card Key Storage Provider"="Microsoft Smart Card Key Storage Provider"
"80000001"="C:\\Program Files\\OpenSC Project\\OpenSC\\minidriver\\opensc-minidriver.dll"
"ATR"=hex:3b,de,18,ff,81,91,fe,1f,c3,80,31,81,54,48,53,4d,31,73,80,21,40,81,07,\
  1c
"ATRMask"=hex:ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,\
  ff,ff,ff
```
"Crypto Provider" gives the name of the provider which could be "Microsoft Base Smart Card Crypto Provider" "Smart Card Key Storage Provider" is sort of a flag to say the minidriver supports CNG i.e. KSP.  

OpenSC also adds the following as an alias for "Microsoft Base Smart Card Crypto Provider" so  one can use `-provider="OpenSC CSP` on some commands. 
 
```
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Defaults\Provider\OpenSC CSP]
"Image Path"="basecsp.dll"
"Type"=dword:00000001
```
Have a look at `[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Defaults\Provider\Microsoft Base Smart Card]` to see how the alias works. It loads `basecsp.dll'  just as the OpenSC CSP does.

**Comment by dengert** [comment]
Disregard the registry issue, it looks like it got far enough to see the card. 

MD virtual file system: found 'mscp' directory
CardReadFile(): file 'msroots' not found in 'mscp' 
It is not required, so this is not an error. 

If that is the whole log it is not clear why it quit. 
It does not look like it tried to read  anything. 

Is this a self signed certificate? 
If so it could be it is looking for the issuer.  You may need to import the certificate into the 
cert store as a trusted CA first.  

You can get more debugging by setting in  `HKEY_LOCAL_MACHINE\SOFTWARE\OpenSC Project\OpenSC` "MiniDriverDebug"=7 and also set in opensc.conf  debug=7; or environment OPENSC_DEBUG=7

**Comment by frankmorgner** [comment]
Yes, the nitrokey contains a smartcard-hsm. does `pkcs11-tool --test --login` work? are you using the changes from https://github.com/OpenSC/OpenSC/pull/1682 and https://github.com/OpenSC/OpenSC/pull/1795?

**Comment by mti-sk** [comment]
Hi,


everything works, but IE not. pkcs12-tool - works, pkcs12-tool - works, openVPN connection via pkcs12-works, pageant-cac via pkcs12 and via CAPI - works.
Neither log minidriver nor openSC shows anything unusual.

For example, there is no problem using Yubukey FIPS, but it does not use opensc-minidriver, but ykmd.dll

I apologize for the foreign language, but I could not (apparently openSC) convince to English.
Chyba čipové karty=Error smartcard,
Tato čipová karta vyžaduje ovladače = This smartcard require driver, which they are not on system.  

![obrazek](https://user-images.githubusercontent.com/7874262/69636159-47705f00-1056-11ea-80ae-010a88a3f74a.png)

**Comment by frankmorgner** [comment]
could you try `certutil -scinfo`?

**Comment by mti-sk** [comment]
Solved in openSC-0.20.0 release
